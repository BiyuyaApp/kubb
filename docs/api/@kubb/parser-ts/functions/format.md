[API](../../../packages.md) / [@kubb/parser-ts](../index.md) / format

# format()

```ts
function format(source): string
```

Format the generated code based on the TypeScript printer.

## Parameters

â€¢ **source**: `string`

## Returns

`string`

## Defined in

[packages/parser-ts/src/format.ts:6](https://github.com/kubb-project/kubb/blob/ff80665146ae086e044807d0072fda660e72e1fd/packages/parser-ts/src/format.ts#L6)
