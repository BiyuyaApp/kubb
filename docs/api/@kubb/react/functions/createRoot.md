[API](../../../packages.md) / [@kubb/react](../index.md) / createRoot

# createRoot()

```ts
function createRoot<Context>(options): ReactTemplate<Context>
```

## Type Parameters

• **Context** *extends* `RootContextProps` = `RootContextProps`

## Parameters

• **options**: `ReactTemplateOptions` = `...`

## Returns

`ReactTemplate`\<`Context`\>

## Defined in

[packages/react/src/createRoot.ts:5](https://github.com/kubb-project/kubb/blob/ff80665146ae086e044807d0072fda660e72e1fd/packages/react/src/createRoot.ts#L5)
