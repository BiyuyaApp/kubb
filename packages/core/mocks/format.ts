import { format as prettierFormat } from 'prettier'
import parserTypescript from 'prettier/parser-typescript.js'

import type { Options } from 'prettier'

const formatOptions: Options = {
  tabWidth: 2,
  printWidth: 160,
  parser: 'typescript',
  singleQuote: true,
  semi: false,
  bracketSameLine: false,
  endOfLine: 'auto',
  plugins: [parserTypescript],
}
export function format(source?: string) {
  if (!source) {
    return ''
  }
  return prettierFormat(source, formatOptions)
}
