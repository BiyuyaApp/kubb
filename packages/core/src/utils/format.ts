import { format as prettierFormat } from 'prettier'

import type { Options } from 'prettier'

const formatOptions: Options = {
  tabWidth: 2,
  printWidth: 160,
  parser: 'typescript',
  singleQuote: true,
  semi: false,
  bracketSameLine: false,
  endOfLine: 'auto',
}
export const format = (text: string) => {
  return prettierFormat(text, formatOptions)
}
