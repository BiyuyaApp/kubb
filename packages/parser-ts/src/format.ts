import { print } from "./print.ts";

export function format(source: string){


  // do some basic linting with the ts compiler
  return print([],{source, noEmitHelpers: false})
}
