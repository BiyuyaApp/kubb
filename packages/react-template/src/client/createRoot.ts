import { createNode } from './dom.ts'
import { ReactTemplate } from './ReactTemplate.tsx'

import type { AppContextProps } from '../components/AppContext.tsx'
import type { DOMElement } from '../types.ts'
import type { RootType } from './types'

const instances = new Map<string, ReactTemplate>()

export function createRoot<Context extends AppContextProps = AppContextProps>(container?: DOMElement): RootType<Context> {
  if (!container) {
    container = createNode('kubb-root')
  }

  const instance = new ReactTemplate<Context>(container)
  instances.set(instance.id, instance)

  return {
    render(children, context?: Context) {
      return instance.render(children, context)
    },
    unmount() {
      instance.unmount()
      instances.delete(instance.id)
    },
    get output() {
      return instance.output
    },
    get imports() {
      return instance.imports
    },
    get exports() {
      return instance.exports
    },
  }
}
