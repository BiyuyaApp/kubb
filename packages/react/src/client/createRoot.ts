import { createNode, ReactTemplate } from '../shared/index.ts'

import type { AppContextProps } from '../components/AppContext.tsx'
import type { DOMElement } from '../types.ts'
import type { RootType } from './types.ts'

const instances = new Map<string, ReactTemplate>()

export function createRoot<Context extends AppContextProps = AppContextProps>(container?: DOMElement): RootType<Context> {
  if (!container) {
    container = createNode('kubb-root')
  }

  const instance = new ReactTemplate<Context>(container)
  instances.set(instance.id, instance)

  return {
    render(children, context?: Context) {
      return instance.render(children, context)
    },
    unmount() {
      instance.unmount()
      instances.delete(instance.id)
    },
    get output() {
      return instance.output
    },
    get imports() {
      return instance.imports
    },
    get exports() {
      return instance.exports
    },
    get file() {
      return instance.file
    },
  }
}
