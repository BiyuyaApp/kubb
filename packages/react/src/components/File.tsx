import { Export } from './Export.tsx'
import { Import } from './Import.tsx'

import type { KubbFile } from '@kubb/core'
import type { KubbNode } from '../types.ts'

type BasePropsWithBaseName = {
  baseName: KubbFile.BaseName
  path: KubbFile.Path
}

type BasePropsWithoutBaseName = {
  baseName?: never
  path?: KubbFile.Path
}

type BaseProps = BasePropsWithBaseName | BasePropsWithoutBaseName

type Props<TMeta extends KubbFile.FileMetaBase = KubbFile.FileMetaBase> = BaseProps & {
  id?: string
  env?: NodeJS.ProcessEnv
  children?: KubbNode
  override?: boolean
  meta?: TMeta
}

export function File<TMeta extends KubbFile.FileMetaBase = KubbFile.FileMetaBase>(props: Props<TMeta>): KubbNode {
  if (!props.baseName || !props.path) {
    return props.children
  }

  return <kubb-file {...props} />
}

type FileSourceUnionProps = {
  /**
   * When path is set it will copy-paste that file as a string inside the component
   * Children will then be ignored
   */
  path?: string
  children?: never
} | {
  /**
   * When path is set it will copy-paste that file as a string inside the component
   * Children will then be ignored
   */
  path?: never
  children?: KubbNode
}

type FileSourceProps = FileSourceUnionProps & {
  print?: boolean
  /**
   * Removes comments
   */
  removeComments?: boolean
  noEmitHelpers?: boolean
}

function FileSource({ path, print, removeComments, noEmitHelpers,children }: FileSourceProps): KubbNode {
  return (
    <kubb-source path={path} print={print} removeComments={removeComments} noEmitHelpers={noEmitHelpers}>
      {children}
    </kubb-source>
  )
}

File.Export = Export
File.Import = Import
File.Source = FileSource
