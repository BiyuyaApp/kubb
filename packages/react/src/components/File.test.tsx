import pathParser from 'node:path'
import path from 'node:path'

import { format } from '../../mocks/format.ts'
import { createRoot } from '../client/createRoot.ts'
import { File } from './File.tsx'

describe('<File/>', () => {
  const mocksPath = path.resolve(__dirname, '../../mocks')

  test('render File', () => {
    const Component = () => {
      return <File fileName="fileName" path="path" />
    }
    const root = createRoot()
    root.render(<Component />)

    expect(root.output).toStrictEqual('')
  })

  test.todo('render File with Import and Export')

  test('render File with source', () => {
    const Component = () => {
      return (
        <File fileName="fileName" path="path">
          <File.Source>test</File.Source>
        </File>
      )
    }
    const root = createRoot()
    root.render(<Component />)

    expect(root.output).toStrictEqual('test')
  })
  test('render File with source and path, print', async () => {
    const Component = () => {
      return (
        <File fileName="fileName" path="path">
          <File.Source path={pathParser.resolve(mocksPath, './test.ts')} print></File.Source>
        </File>
      )
    }
    const root = createRoot()
    root.render(<Component />)

    expect(format(root.output)).toStrictEqual(
      format(`export function test() {
      return true
    }`),
    )
  })
})
