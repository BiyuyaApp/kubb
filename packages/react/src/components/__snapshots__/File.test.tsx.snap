// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<File/> > render File with Import and Export and print imports/exports 1`] = `
"import React from 'react'
export * from './index.ts'
"
`;

exports[`<File/> > render File with source and path, print 1`] = `Promise {}`;
